### ISSUE TITLE
`--environment <docker|systemd>` takes precedence over "logging.to_files: true"

### ISSUE THREAD

**Issue by martinmorch-rsyd** [elastic/beats#41767]
## Edit

Since https://github.com/elastic/beats/pull/41186 got merged, `--environment <docker|systemd>` is taking precedence over `logging.to_files: true`. The old behaviour was for the logging configuration in the configuration file to take precedence over the `--environment` flag.

For reference, `v8.14.3` is the latest release with the correct behaviour of the `--environment` CLI argument.

More details on https://github.com/elastic/beats/issues/41767#issuecomment-2518584937

## Original post

Filebeat versions of 8.15.4 and higher no longer logs to "logging.files.path" when "logging.to_files: true" is set, unless "logging.to_stderr: false" is explicitly set.

In versions <=8.15.3, setting "logging.to_stderr: false" was not required for "logging.to_files: true" to be honored.

- Filebeat >=8.15.4: 
- Operating System: Rocky Linux 8.10, Ubuntu 22.04.5
- Steps to Reproduce: 

1. Run Filebeat version >=8.15.4 with the following logging configuration:
```
logging.level: info
logging.to_files: true
logging.files:
  path: /var/log/filebeat
  name: filebeat
  keepfiles: 7
  permissions: 0640
```
2. Log files are not created under "logging.file.path".
3. Revert to running Filebeat <=8.15.3 with the same configuration.
4. Log files are created under "logging.file.path".

I see nothing in the release notes about this behavioral change, is it expected?

**Comment by amardeep-s** [comment]
I am facing the same issue on all nodes where Filebeat was upgraded. 

- Filebeat >=8.15.4
- Operating System: Rocky Linux 8.10, Rocky Linux 9

Spend hours to troubleshoot and trying different combinations, despite no changes being made in filebeat.yml before upgrading filebeat.

The current workaround is downgrade to filebeat-8.13.2 which is working as expected, or change **logging.level: error** instead of info.

**Comment by belimawr** [comment]
Could you folks share the output of `ps aux |grep filebeat` when running Filebeat?

**Comment by martinmorch-rsyd** [comment]
@belimawr 
```
$ filebeat version
filebeat version 8.16.1 (amd64), libbeat 8.16.1 [f17e0828f1de9f1a256d3f520324fa6da53daee5 built 2024-11-14 14:57:13 +0000 UTC]
$ cat /etc/redhat-release
Rocky Linux release 8.10 (Green Obsidian)
```
```
$ ps aux | grep filebeat
root     3971550  9.7  0.2 2025520 79332 ?       Ssl  07:32   0:00 /usr/share/filebeat/bin/filebeat --environment systemd -c /etc/filebeat/filebeat.yml --path.home /usr/share/filebeat --path.config /etc/filebeat --path.data /var/lib/filebeat --path.logs /var/log/filebeat
```
filebeat.yml:
```
logging:
  level: info
  to_files: true
  to_syslog: false
  files:
    path: /var/log/filebeat
    name: filebeat.log
    keepfiles: 3
    permissions: 0644
```
```
$ ls -la /var/log/filebeat/
total 8
drwxr-xr-x.  2 root root    6 Dec  4 07:35 .
drwxr-xr-x. 17 root root 4096 Dec  2 15:00 ..
```

**Comment by belimawr** [comment]
Thanks for the information folks!

So the CLI argument `--environment systemd` is affecting the logging behaviour. We use `--environment` to tell Filebeat (well, any Beat) the environment they're running at so they can adjust their behaviour accordingly, which includes logging. There are a few environments we default to logging to stderr: `docker`, `systemd`, etc.

Some updates on our logging introduced a bug on `v8.15.2` (https://github.com/elastic/beats/issues/41118) that made Beats to not log to stderr when running on docker, systemd, etc. It got fixed recently (https://github.com/elastic/elastic-agent-libs/pull/236) and released in `v8.15.4`, however we missed the changelog entry 🤦‍♂ and now `--environment [systemd|docker]` is taking precedence over setting `logging.to_files: true` in the configuration YAML 🤦‍♂ 🤦‍♂

**Comment by belimawr** [comment]
As a wokaround, you'll need to set `logging.to_stderr: false`, which disables logging to stderr and set whichever other output you want to `true`.

```yaml
logging.level: info
logging.to_files: true
logging.to_stderr: false
```

I'll edit the original post to be a bit more clear about it.

**Comment by redbaron4** [comment]
@belimawr Does this mean that from now on `--environment systemd` will take precedence over logging configuration in config file? In that case docs will have to be updated because currently the doc page for command line options says ```
This setting is used to select a default log output when no log output is configured ```

**Comment by belimawr** [comment]
> [@belimawr](https://github.com/belimawr) Does this mean that from now on `--environment systemd` will take precedence over logging configuration in config file? In that case docs will have to be updated because currently the doc page for command line options says `This setting is used to select a default log output when no log output is configured`

No, the precedence order should not change, we need to fix the implementation.

**Comment by swtrux** [comment]
What version can we expect a fix for this to be in?  This will cause a significant increase in logging to /var/log/messages on our system and we would like to wait for the fix before upgrading.
